pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java

```diff
@@ -0,0 +1,373 @@
+     */
+    private void checkAssignExceptIncrement(Object data, Set<String> loopVariables, ASTStatement loopBody, IgnoreFlags... ignoreFlags) {
+        checkAssignments(data, loopVariables, loopBody, false, ignoreFlags);
+    }
+
+    /**
+     * Report usages of increments ('++', '--', '+=', '-=').
+     *
+     * @param ignoreFlags which statements should be ignored
+     */
+    private void checkIncrementAndDecrement(Object data, Set<String> loopVariables, ASTStatement loopBody, IgnoreFlags... ignoreFlags) {
+
+        // foo ++ and foo --
+        for (ASTPostfixExpression expression : loopBody.findDescendantsOfType(ASTPostfixExpression.class)) {
+            if (ignoreNode(expression, loopBody, ignoreFlags)) {
+                continue;
+            }
+
+            checkVariable(data, loopVariables, expression.getFirstDescendantOfType(ASTName.class));
+        }
+
+        // ++ foo
+        for (ASTPreIncrementExpression expression : loopBody.findDescendantsOfType(ASTPreIncrementExpression.class)) {
+            if (ignoreNode(expression, loopBody, ignoreFlags)) {
+                continue;
+            }
+
+            checkVariable(data, loopVariables, expression.getFirstDescendantOfType(ASTName.class));
+        }
+
+        // -- foo
+        for (ASTPreDecrementExpression expression : loopBody.findDescendantsOfType(ASTPreDecrementExpression.class)) {
+            if (ignoreNode(expression, loopBody, ignoreFlags)) {
+                continue;
+            }
+
+            checkVariable(data, loopVariables, expression.getFirstDescendantOfType(ASTName.class));
+        }
+
+        // foo += x and foo -= x
+        checkAssignments(data, loopVariables, loopBody, true, ignoreFlags);
+    }
+
+    /**
+     * Report usages of assignments.
+     *
+     * @param checkIncrement true: check only '+=' and '-=',
+     *                       false: check all other assignments
+     * @param ignoreFlags    which statements should be ignored
+     */
+    private void checkAssignments(Object data, Set<String> loopVariables, ASTStatement loopBody, boolean checkIncrement, IgnoreFlags... ignoreFlags) {
+        for (ASTAssignmentOperator operator : loopBody.findDescendantsOfType(ASTAssignmentOperator.class)) {
+            // check if the current operator is an assign-increment or assign-decrement operator
+            final String operatorImage = operator.getImage();
+            final boolean isIncrement = "+=".equals(operatorImage) || "-=".equals(operatorImage);
+
+            if (isIncrement != checkIncrement) {
+                // wrong type of operator
+                continue;
+            }
+
+            if (ignoreNode(operator, loopBody, ignoreFlags)) {
+                continue;
+            }
+
+            final ASTPrimaryExpression primaryExpression = operator.jjtGetParent().getFirstChildOfType(ASTPrimaryExpression.class);
+            final ASTName name = primaryExpression.getFirstDescendantOfType(ASTName.class);
```

Comment:
Looking through the AST: By traversing through `ASTPrimaryExpression -> ASTPrimaryPrefix -> ASTName` I can make sure that I have variable's name (or null, if the expression is `this.` or `super.`)
Additionally checking for `ASTPrimaryExpression -> ASTPrimarySuffix` allows detecting arrays.
This method doesn't interfere with detecting `array[i = 4] =...`: It's treated the same way as `doSomethingWith(i = 4)`
.
I've added these checks and additional test cases.


But I'm a little bit confused by the representations of different kinds of field accesses in the AST.
Given this code:
```java
public class Foo {
    void foo() {
        for(String[] bar : bars()) {
            bar = new String[10]; // this is the only actual reassignment of bar
            bar[0] = "foo";
            bar.length = 5; // let's assume that array.length is writable...
            this.bar = 5;
        }
    }
}
```
The four assignments are parsed as:
- `PrimaryPrefix/Name[@Image='bar']`
- `PrimaryPrefix/Name[@Image='bar']` and `PrimarySuffix[@ArrayDereference=true]`
- `PrimaryPrefix/Name[@Image='bar.length']`
- `PrimaryPrefix[@ThisModifier=true]` and `PrimarySuffix[@Image='bar']`

Shouldn't the third one be parsed as `PrimaryPrefix/Name[@Image='bar']` and `PrimarySuffix[@Image='length']`?
I'm not sure if this is a bug or an intended feature/behavior...

Reply:
Not sure, whether that's a bug or not. It probably just evolved in that way and now it is that way - this/super probably have been integrated later into the grammar. I've added a note to #1019 as changing this would break a lot of things... so for now we need to live with it, but we can change it for 7.0.0.